server/src/catalog.rs

```diff
@@ -314,22 +323,67 @@ impl NewModelTableMetadata {
             return Err("The timestamp column cannot be a tag column.".to_owned());
         };
 
-        // If the index of the timestamp column does not match the schema, return an error.
-        if schema.fields.get(timestamp_column_index as usize).is_none() {
-            return Err("The timestamp column index does not match a field in the schema.".to_owned());
-        };
+        if let Some(timestamp_field) = schema.fields.get(timestamp_column_index as usize) {
+            // If the field of the timestamp column is not of type ArrowTimestamp, return an error.
+            if !timestamp_field.data_type().equals_datatype(&ArrowTimestamp::DATA_TYPE) {
+                return Err("The timestamp column does not have the correct data type.".to_owned());
+            }
+        } else {
+            // If the index of the timestamp column does not match the schema, return an error.
+            return Err(
+                "The timestamp column index does not match a field in the schema.".to_owned(),
+            );
+        }
 
-        // If the indices for the tag columns does not match the schema, return an error.
         for tag_column_index in &tag_column_indices {
-            if schema.fields.get(*tag_column_index as usize).is_none() {
-                return Err(format!("The tag column index '{}' does not match a field in the schema.", tag_column_index));
-            };
-        };
+            if let Some(tag_field) = schema.fields.get(*tag_column_index as usize) {
+                // If the fields of the tag columns is not of type Utf8, return an error.
+                if !tag_field.data_type().equals_datatype(&DataType::Utf8) {
+                    return Err(format!(
+                        "The tag column with index '{}' does not have the correct data type.",
+                        tag_column_index
+                    ));
+                }
+            } else {
+                // If the indices for the tag columns does not match the schema, return an error.
+                return Err(format!(
+                    "The tag column index '{}' does not match a field in the schema.",
+                    tag_column_index
+                ));
+            }
+        }
+
+        let field_column_indices: Vec<usize> = (0..schema.fields().len()).filter(|index| {
+            let index = *index as u8;
```

Comment:
Since this is something that will be changed in the future, I have just added a TODO to change the cast when the indices in the action body are parsed differently.

Reply:
Ok @CGodiksen , if we have to significantly extend the current action it may be simpler to write a quick scanner and parser that can process the simple `CREATE TABLE` SQL statements we require for tables and model tables until [Apache DataFusion's parser can be extended](https://github.com/apache/arrow-datafusion/blob/master/docs/source/contributor-guide/roadmap.md). This would probably also allow us to create a simpler user-interface, e.g., an action that simply accepts a string like `CREATE TABLE normal_table(id INTEGER, name...)` (although maybe we can use [Apache DataFusion's existing support for DDL](https://arrow.apache.org/datafusion/user-guide/sql/ddl.html)) and `CREATE TABLE model_table(timestamp TIMESTAMP, field_1 FIELD(10%), field_2 FIELD(0%), tag TAG)` (although I do not know if `FIELD(0%)` deviates too much from existing SQL convention `VARCHAR(10)`).

Reply:
I think having a custom parser is a good idea for the future. Communicating to the user how to put together a single list of bytes to provide the arguments seems a lot harder than just letting them use SQL.